Basic overloading

beta-ziliani opened the pull request:
Lots of TODOs pending

beta-ziliani committed:
initial implementation of overloading

beta-ziliani committed:
Adding implicit casting

beta-ziliani committed:
Adding examples

beta-ziliani committed:
fixup

beta-ziliani committed:
Solving this and super

beta-ziliani committed:
Negative case of ambiguous overload (make sure it won't resolve to an arbitrary ref)

beta-ziliani committed:
Handling of call options

beta-ziliani committed:
Adding test cases

ggiraldez committed:
Refactor function call expression typing

ggiraldez committed:
Fix tests

ggiraldez committed:
Select first candidate when resolving overloads from functions

If after filtering functions for matching parameter types there are more than
one candidate left, pick the first one. Multiple remaining candidates are very
likely because of overriden virtual methods, as the resolution algorithm will
return all available functions that match the given name.

ggiraldez committed:
Add test cases for overloaded attached functions

ggiraldez committed:
Matching overloads, consider the receiver as an implicit first argument

ggiraldez committed:
Add test case for call with multiple overloads with named arguments

ggiraldez committed:
Select function overloads when invoked with named arguments

ggiraldez committed:
Resolve ambiguity of constants and state variables by picking the first

In valid Solidity code, if we find other definitions they are either overriden
or private.

ggiraldez committed:
Legacy options for `staticcall` and `delegatecall`

ggiraldez committed:
Relax visibility requirements for functions in Solidity < 0.5.0

ggiraldez committed:
Fixup typing of `var` declarations

ggiraldez committed:
Don't register legacy constructors in the contract scope

They cannot be invoked in any way (except through a normal `new`) and having
them in scope leads to ambiguity when resolving the contract type.

beta-ziliani commented on crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:
```diff
@@ -829,11 +914,41 @@ impl Pass {
+            Typing::Undetermined(type_ids) => {
                 // TODO: resolve argument types and match best overload
                 // TODO: maybe update the typing of the operand?
                 // TODO: return the definition_id used to later resolve named arguments
```
These todo's should be removed, don't they? Same for the positional arguments.

beta-ziliani reviewed (commented):
Awesome! Just left a comment about the todo's.

ggiraldez committed:
Fix overload disambiguation when `this` is an argument

Because `this` typing is `Typing::This` it was not being properly treated as an
address type.

ggiraldez committed:
Member access expression typing inherits data location of operand

ggiraldez committed:
In external calls all data locations are implicitly convertible to memory and calldata

ggiraldez committed:
Implicit conversions of reference types for function calls

ggiraldez committed:
Try to resolve attached functions before built-ins

ggiraldez committed:
Filter attached functions by the receiver type

Also add implicit conversion from a contract/interface to its bases

ggiraldez committed:
Fix typing of `bytes` index access and range access for other types

ggiraldez committed:
Type indexing of byte arrays

ggiraldez committed:
Support typing `abi.decode` built-in

ggiraldez committed:
When linearising contracts, resolve only contracts and interfaces in the file scope

beta-ziliani commented on crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs:
```diff
@@ -542,9 +543,12 @@ impl<'a> BuiltInsResolver<'a> {
+                if *external || self.language_version < VERSION_0_5_0 {
                     match symbol {
-                        "address" => Some(BuiltIn::Address),
+                        "address" if self.language_version >= VERSION_0_8_2 => {
```
I was expecting this condition to appear in the tests below, but it's being swallowed by the visibility issue (IIUC). Perhaps it's worth to break the tests in two, one testing the visibility, and one testing the address?

beta-ziliani replied:
Or is this new functionality and can't be tested in today's setting?

ggiraldez replied:
I don't remember, but this is probably ported over from the built-ins defined in the language definition (see https://github.com/manastech/slang/blob/main/crates/solidity/inputs/language/src/definition.rs#L6682)

beta-ziliani commented on crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:
```diff
@@ -1608,4 +1608,53 @@ impl Visitor for Pass {
+        let types = if let Type::Tuple { types } = self.types.get_type_by_id(tuple_type_id) {
+            types.as_slice()
+        } else {
+            // the resolved type is not a tuple
```
Isn't this a logic error? or in which scenario we're leaving a tuple deconstruction of something that's not a tuple?

ggiraldez replied:
In general, a single element tuple is collapsed to a single value. In this case though it's the opposite, because because the typing of the expression may not yield a tuple. But we still want to handle cases like:

```solidity
(uint x) = foo();
```
Which is parsed as a tuple deconstruction statement.

beta-ziliani commented on crates/solidity/testing/snapshots/bindings_output/variables/var_type_detection/input.sol:
none of the tests succeeds. Is this yet another thing in which the new binder is better?

ggiraldez replied:
Yes, indeed. This could only pass in < 0.5.0 (because of the deprecated `var` keyword), but the old binder cannot assign the resulting type of the initializer expression to the variable. Then, member accesses to the variables fail to resolve.

beta-ziliani commented on crates/solidity/testing/snapshots/bindings_output/contracts/inherited_legacy_constructor/generated/0.4.11-success.txt:
```diff
@@ -0,0 +1,71 @@
+    │                               ╰──── name: 6
+  9 │         token old = token(address(0x7F2176cEB16dcb648dc924eff617c3dC2BEfd30d));
+    │         ──┬── ─┬─   ──┬──  
+    │           ╰──────────────── ref: 3
```
The new binder returns `ref: 1` here, I presume?

ggiraldez replied:
Yes!

ggiraldez committed:
Resolve ambiguity in event overloads

ggiraldez committed:
Type `msg.data` as `bytes calldata` instead of `bytes memory`

beta-ziliani commented on crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:
```diff
@@ -334,6 +334,38 @@ impl Pass {
+            return false;
+        };
+        let Type::Function(function_type) = self.types.get_type_by_id(definition_type_id) else {
+            // the definition is not a function
```
More like the definition doesn't have the right type? Because it _is_ a function, according to the first check. Or am I missing something?

ggiraldez replied:
Hmm... what the above checks is that the _definition_ should be a function, whereas this checks the type bound to that definition. I guess this would be a logic error and better served by an `unreachable!`.

beta-ziliani reviewed (commented):
I keep tripping with the (old binder) tests 😅 IIUC the new binder is handling many situations that the old one is unable to 💯

ggiraldez committed:
When a resolution is ambiguous, filter out overriden definitions

This naturally applies to function definitions, but also in the case of public
state variables that override a method in a base clase with the same type
signature.

ggiraldez committed:
Use implicit conversion rules for external calls when filtering attached functions

ggiraldez committed:
Fix getter type computation for structs

ggiraldez committed:
Consider false branch when typing a conditional expression

ggiraldez committed:
Revamp typing for literals, mostly to add implicit conversion cases

We still need to improve the handling of numbers in general, as solc will
normally reify a literal number to the smallest possible elementary type it can.

ggiraldez committed:
Add support for old `DUPi` and `SWAPi` Yul opcodes in < 0.5.0

ggiraldez committed:
Fix implicit conversions between literal numbers

ggiraldez committed:
Some minor updates from code review

ggiraldez committed:
Fix scoping rules for declaration statements

Variables defined in declaration statements (variables and tuple deconstruction)
are only available *after* the statement.

ggiraldez commented:
Closing this, since I submitted https://github.com/NomicFoundation/slang/pull/1426 with this and some other commits.
